src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsUrlTest.java

```diff
@@ -0,0 +1,189 @@
+                        "module name: %s is not exist in %s", moduleName, PACKAGE_NAME);
+                assertWithMessage(moduleNameErrorMsg).that(moduleFileNames).isNotNull();
+                boolean match = false;
+                final String checkNameWithSuffix = checkNameInAttribute + SUFFIX_CHECK_JAVA;
+                for (String filename : moduleFileNames) {
+                    if (checkNameWithSuffix.equals(filename)) {
+                        match = true;
+                        break;
+                    }
+                }
+                assertWithMessage(checkNameModuleErrorMsg).that(match).isTrue();
+            }
+        }
+    }
+
+    public static final class CheckTest extends DefaultHandler {
+
+        public static final String SPLIT_CHECK_NAME_IN_ATTRIBUTE = "#";
+
+        public static final String SPLIT_MODULE_NAME_IN_ATTRIBUTE = "\\.";
+
+        public static final String PREFIX_CONFIG = "config_";
+
+        private static final String NODE_NAME = "a";
+
+        private List<List<String>> checkNamesList;
+
+        private List<String> singleCheckNameList;
+
+        private String currentTag;
+
+        @Override
+        public void startDocument() {
+            checkNamesList = new ArrayList<>();
+        }
+
+        @Override
+        public void startElement(String uri, String localName, String qName,
+                                 Attributes attributes) {
+            if (qName.equals(NODE_NAME)) {
+                final String[] moduleAndCheckName =
+                        attributes.getValue(0).split(SPLIT_CHECK_NAME_IN_ATTRIBUTE);
+                if (moduleAndCheckName[0].startsWith(PREFIX_CONFIG)) {
+                    singleCheckNameList = new ArrayList<>();
+                    final String moduleNameWithHtml =
+                            moduleAndCheckName[0].substring(PREFIX_CONFIG.length());
+                    final String moduleName =
+                            moduleNameWithHtml.split(SPLIT_MODULE_NAME_IN_ATTRIBUTE)[0];
+                    singleCheckNameList.add(moduleName);
+                    singleCheckNameList.add(moduleAndCheckName[1]);
+                }
+            }
+            currentTag = qName;
+        }
+
+        @Override
+        public void characters(char[] ch, int start, int length) {
+            if (currentTag != null && singleCheckNameList != null && currentTag.equals(NODE_NAME)) {
+                final String currentValue = new String(ch, start, length).trim();
+                if (!currentValue.isEmpty() && !"\n".equals(currentValue)) {
+                    singleCheckNameList.add(currentValue);
+                }
+                currentTag = null;
+            }
+        }
+
+        @Override
+        public void endElement(String uri, String localName, String qName) {
+            if (singleCheckNameList != null && qName.equals(NODE_NAME)) {
+                checkNamesList.add(singleCheckNameList);
+                singleCheckNameList = null;
```

Comment:
it is better to have this list and clear it every time, e.g.
```
if(qName.equals(NODE_NAME)) {
  checkNamesList.addAll(singleCheckNameList);
  singleCheckNameList.clear();
}
```
And you can remove line 158 and init field with empty list
Same can be done for checkNamesList ( can be cleared in endDocument method)

Reply:
it is not proper, although singleCheckNameList is added to checkNamesList, two references point to the same object, so if clear(), checkNamesList will contain nothing. Same as checkNamesList if it is cleared in endDocument, error will happen in "for (List<String> sub : checkHandler.checkNamesList) {" which is empty